subsys/net/ip/net_context.c

```diff
@@ -1175,40 +1211,38 @@ NET_CONN_CB(tcp_established)
 		return NET_DROP;
 	}
 
-	ret = packet_received(conn, pkt, context->tcp->recv_user_data);
+	/* If the pkt with appdata, notify the recv callback which should
+	 * release the pkt if handled. If not, release the pkt immediately at
+	 * clean_up.
+	 */
+	if (data_len > 0) {
+		ret = packet_received(conn, pkt, context->tcp->recv_user_data);
+	}
 
+	/* Increment the ack */
 	context->tcp->send_ack += data_len;
-
 	if (tcp_flags & NET_TCP_FIN) {
-		/* Sending an ACK in the CLOSE_WAIT state will transition to
-		 * LAST_ACK state
-		 */
-		context->tcp->fin_rcvd = 1;
+		context->tcp->send_ack += 1;
+	}
 
-		if (net_tcp_get_state(context->tcp) == NET_TCP_ESTABLISHED) {
-			net_tcp_change_state(context->tcp, NET_TCP_CLOSE_WAIT);
-		}
+	send_ack(context, &conn->remote_addr, false);
 
-		context->tcp->send_ack += 1;
+clean_up:
+	if (net_pkt_appdatalen(pkt) == 0) {
+		net_pkt_unref(pkt);
```

Comment:
This unref is a bit problematic as I am seeing warnings about double unref. Application will release the net_pkt when it receives it, and if the tcp_established() returns NET_DROP, then the packet is also freed in net_core.c:processing_data().
Was this unref here so that if we receive final ack, we return NET_OK and there would be memory leak otherwise?
If so, then better would be to put the net_pkt_unref() where you call clean_up.

In order to avoid the double unref, I used `if (net_pkt_appdatalen(pkt) == 0 && ret == NET_OK) {`  here but there were still some unref warnings as the application might have released the net_pkt already.

I tested this with echo-client with CONFIG_NET_DEBUG_NET_PKT=y (in order to see memory allocations), and #4972 applied.  It seemed to work but at some point the echo-client hang and stopped responding, I am trying to find out is this patch causing that or is there some unrelated issue.

If I remove this unref in this line, then echo-server runs out of memory, but with this unref in place, the echo-server runs just fine and I do not see any issues.

Reply:
Indeed....it will cause double unref because the pkt is released by the application and appdata len is unpredictable. I will try to reproduce the issue by echo client as you mentioned.

Is it a good idea to copy the appdata len to data_len at the begining of tcp_establish()? So we can use data len to decide whether we should unref the pkt in tcp_establish() or not.

Reply:
Btw, echo_server crash I quote in https://github.com/zephyrproject-rtos/zephyr/pull/5015#issuecomment-345198876 happens in net_context_unref(), so is likely the same issue.

Reply:
> Was this unref here so that if we receive final ack, we return NET_OK and there would be memory leak otherwise?

@jukkar, Yes. I also want to put the pkt unref together for clearness since it was hard for me to trace the pkt lifecycle.
However, to prevent double unref, I decide to remove pkt unref from clean_up.